proto/src/control.rs

```diff
@@ -0,0 +1,613 @@
+                state,
+                entry_uuid,
+                cookie,
+            } => {
+                let d_cookie = cookie.as_ref().map(|s| bytes_to_string!(s));
+                f.debug_struct("LdapControl::SyncState")
+                    .field("state", &state)
+                    .field("entry_uuid", &entry_uuid)
+                    .field("cookie", &d_cookie)
+                    .finish()
+            }
+            LdapControl::SyncDone {
+                cookie,
+                refresh_deletes,
+            } => {
+                let d_cookie = cookie.as_ref().map(|s| bytes_to_string!(s));
+                f.debug_struct("LdapControl::SyncDone")
+                    .field("refresh_deletes", &refresh_deletes)
+                    .field("cookie", &d_cookie)
+                    .finish()
+            }
+            LdapControl::AdDirsync {
+                flags,
+                max_bytes,
+                cookie,
+            } => {
+                let d_cookie = cookie.as_ref().map(|s| bytes_to_string!(s));
+                f.debug_struct("LdapControl::AdDirsync")
+                    .field("flags", &flags)
+                    .field("max_bytes", &max_bytes)
+                    .field("cookie", &d_cookie)
+                    .finish()
+            }
+            LdapControl::SimplePagedResults { size, cookie } => {
+                let d_cookie = bytes_to_string!(cookie);
+                f.debug_struct("LdapControl::SimplePagedResults")
+                    .field("size", &size)
+                    .field("cookie", &d_cookie)
+                    .finish()
+            }
+            LdapControl::ManageDsaIT { criticality } => f
+                .debug_struct("LdapControl::ManageDsaIT")
+                .field("criticality", &criticality)
+                .finish(),
+        }
+    }
+}
+
+impl TryFrom<StructureTag> for LdapControl {
+    type Error = LdapProtoError;
+
+    fn try_from(value: StructureTag) -> Result<Self, Self::Error> {
+        let mut seq = value
+            .match_id(Types::Sequence as u64)
+            .and_then(|t| t.expect_constructed())
+            .ok_or(LdapProtoError::ControlBer)?;
+
+        // We destructure in reverse order due to how vec in rust
+        // works.
+        let (oid_tag, criticality_tag, value_tag) = match seq.len() {
+            1 => {
+                let v = None;
+                let c = None;
+                let o = seq.pop();
+                (o, c, v)
+            }
+            2 => {
+                let v = seq.pop();
+                let c = None;
+                let o = seq.pop();
+                (o, c, v)
+            }
+            3 => {
+                let v = seq.pop();
+                let c = seq.pop();
+                let o = seq.pop();
+                (o, c, v)
+            }
+            _ => return Err(LdapProtoError::ControlSeqLen),
+        };
+
+        // trace!(?oid_tag, ?criticality_tag, ?value_tag);

```

Comment:
Nitpick: either uncomment or delete this line

Reply:
updated